include/deal.II/particles/particle_accessor.h

```diff
@@ -0,0 +1,210 @@
+
+    /**
+      * Returns whether this particle has a valid property pool and a valid
+      * handle to properties.
+      */
+    bool
+    has_properties () const;
+
+    /**
+     * Set the properties of this particle.
+     *
+     * @param [in] new_properties A vector containing the
+     * new properties for this particle.
+     */
+    void
+    set_properties (const std::vector<double> &new_properties);
+
+    /**
+     * Get write-access to properties of this particle.
+     *
+     * @return An ArrayView of the properties of this particle.
+     */
+    const ArrayView<double>
+    get_properties ();
+
+    /**
+     * Get read-access to properties of this particle.
+     *
+     * @return An ArrayView of the properties of this particle.
+     */
+    const ArrayView<const double>
+    get_properties () const;
+
+    /**
+     * Returns the size in bytes this particle occupies if all of its data is
+     * serialized (i.e. the number of bytes that is written by the write_data
+     * function of this class).
+     */
+    std::size_t
+    serialized_size_in_bytes() const;
+
+    /**
+     * Get a cell iterator to the cell surrounding the current particle.
+     * As particles are organized in the structure of a triangulation,
+     * but the triangulation itself is not stored in the particle this
+     * operation requires a reference to the triangulation.
+     */
+    typename parallel::distributed::Triangulation<dim,spacedim>::cell_iterator
+    get_surrounding_cell (const parallel::distributed::Triangulation<dim,spacedim> &triangulation) const;
+
+    /**
+     * Serialize the contents of this class.
+     */
+    template <class Archive>
+    void serialize (Archive &ar, const unsigned int version);
+
+    /**
+     * Advance the ParticleAccessor to the next particle.
+     */
+    void next ();
+
+    /**
+     * Move the ParticleAccessor to the previous particle.
+     */
+    void prev ();
+
+    /**
+     * Inequality operator.
+     */
+    bool operator != (const ParticleAccessor<dim,spacedim> &other) const;
+
+    /**
+     * Equality operator.
+     */
+    bool operator == (const ParticleAccessor<dim,spacedim> &other) const;
+
+  protected:
+    /**
+     * Construct an invalid accessor. Such an object is not usable.
+     */
+    ParticleAccessor ();
+
+    /**
+     * Construct an accessor from a reference to a map and an iterator to the map.
+     * This constructor is protected so that it can only be accessed by friend
+     * classes.
+     */
+    ParticleAccessor (const std::multimap<types::LevelInd, Particle<dim,spacedim> > &map,
+                      const typename std::multimap<types::LevelInd, Particle<dim,spacedim> >::iterator &particle);
+
+  private:
+    /**
+     * A pointer to the container that stores the particles. Obviously,
+     * this accessor is invalidated if the container changes.
+     */
+    std::multimap<types::LevelInd, Particle<dim,spacedim> >            *map;
```

Comment:
Do we ever want to allow inheriting from the `Particle` class? Currently the `Particle` class has templated `save()` and `load()` functions which could make it tricky to inherit from. However, if we do want to inherit from `Particle`, we may want this class to be templated with `ParticleType`? as in 
```
template<int dim, int spacedim=dim, typename ParticleType=Particle<dim, spacedim>>
class ParticleAccessor ...
```

Reply:
to elaborate on @vishalkenchan comment: it would be good to allow users to extend particles, say in application we work each particle needs to store additionally couple of `Point<dim,spacedim>` and several integer properties. This should work OOB with
```
template<...>
class CustomParticle : public dealii::Particle<dim,spacedim>
{
...
}
```